src/charm.py

```diff
@@ -208,7 +208,7 @@ def __init__(self, *args):
             path=f"{external_url.path}/loki/api/v1/push",
         )
 
-        self.v12_migration_date = self._get_v12_migration_date()
+        self.v12_migration_date = self._get_v12_migration_date
```

Comment:
I've marked some of the other comments as argument below is relevant to all

The reason for keeping this property separate is to determine the value based on the specific hook that fired. The `self._get_v12_migration_date()` job is to get the date from the persisted file. If the file isn't there (i.e., during an upgrade/install), then the event listeners will take care of setting the value. Dropping this in favor of 1 property can be achieved but felt rather complex as we won't have knowledge inside the `self._get_v12_migration_date()` of which hook caused it to execute which will leave us with a couple of options

- always set the migration date to tomorrow (This will lead to complications later on when boltdb is completely removed)
- query the loki api for how many chunks are inserted, 0 chunks meaning this is a fresh install and we can set the TSDB date as today, now the problem with this approach is that it will involve some unnecessary string manipulation for the api response and will enforce us to guard against HTTP and connection errors.